link far end aligns with the mouse pointer position

antonioru opened the pull request:
# Checklist

- [x] The code has been run through pretty `yarn run pretty`
- [x] The tests pass on CircleCI
- [x] You have referenced the issue(s) or other PR(s) this fixes/relates-to
- [x] The PR Template has been filled out (see below)
- [x] Had a beer/coffee because you are awesome

## What?
The `DragNewLinkState.fireMouseMoved` method was not taking into account the click starting position when dragging a new link, it just assumes it's the same point (x,y) of the link port. Sometimes, especially with big ports, it causes a slight mismatch between the link far end point and the mouse position as reported in the issue #467 

<img src="https://i.postimg.cc/bwnwCnmS/Screenshot-2019-11-17-at-23-47-39.png">


## How?

I saved the mouse starting position within the `DragNewLinkState` class itself. 

## Feel good image:

(Add your own one below :])

![LOL](https://i.pinimg.com/originals/7f/1b/c3/7f1bc3fb2e123dd3255a85c04db22f19.jpg)

antonioru commented:
@dylanvorster running `yarn run pretty` will "prettify" some files not related to the Pull Request, I can revert em to the previous state if needed

dylanvorster commented on packages/react-diagrams-core/src/states/DragNewLinkState.ts:
```diff
@@ -56,6 +60,9 @@ export class DragNewLinkState extends AbstractDisplacementState<DiagramEngine> {
 					this.port.reportPosition();
+
+					// save the mouse position for further precision in calculating the link's far-end point
+					this.startingPoint = new Point(event.event.clientX, event.event.clientY);
```
The AbstractDisplacementState already contains this data in the abstract class :)

antonioru replied:
> The AbstractDisplacementState already contains this data in the abstract class :)

that's brilliant! 
I'm going to update the PR, thank you for the tip ;)

antonioru replied:
@dylanvorster MR updated! ;)

mathiasbaldissera commented:
I think it's not working 100%.
I Updated my personal State with your calcs and this became the result
![React App - Chromium_003](https://user-images.githubusercontent.com/22898471/69105490-2c945e00-0a4a-11ea-8803-dc4699c255f8.png)

antonioru commented:
> I think it's not working 100%.
> I Updated my personal State with your calcs and this became the result
> ![React App - Chromium_003](https://user-images.githubusercontent.com/22898471/69105490-2c945e00-0a4a-11ea-8803-dc4699c255f8.png)

Thank you mathiasbaldissera for checking it out, lately I've updated the code by using the internal variables (inernalX, internalY), so the method looks like this now:

```
fireMouseMoved(event: AbstractDisplacementStateEvent): any {
		const pos = this.port.getPosition();
		this.link.getLastPoint().setPosition(pos.x + event.virtualDisplacementX, pos.y + event.virtualDisplacementY);
		const linkNextPosX = pos.x + (this.initialX - pos.x) + event.virtualDisplacementX;
		const linkNextPosY = pos.y + (this.initialY - pos.y) + event.virtualDisplacementY;

		this.link.getLastPoint().setPosition(linkNextPosX, linkNextPosY);
		this.engine.repaintCanvas();
	}
```

can you please confirm this is not working for you?

antonioru commented:
Ok I've better looked into it (and I apologise I haven't done it before) @mathiasbaldissera is right, apparently it works most of the time:

<img src="https://ibb.co/q1QdNXs />

but sometimes it doesn't: 

<img src="https://ibb.co/5W0WYRt" />

I can't really understand why and I'll try to figure out as soon as possible... 

any suggestion @dylanvorster ?

mathiasbaldissera commented:
Using the initialXRelative and initialYRelative less the canvas offset worked to me
```typescript
		const pos = this.port.getPosition()
		const linkNextPosX =
			pos.x + (this.initialXRelative - this.engine.getModel().getOffsetX() - pos.x) + event.virtualDisplacementX
		const linkNextPosY =
			pos.y + (this.initialYRelative - this.engine.getModel().getOffsetY() - pos.y) + event.virtualDisplacementY

		this.link.getLastPoint().setPosition(linkNextPosX, linkNextPosY)
```

antonioru commented:
> ```ts
> const pos = this.port.getPosition()
> 		const linkNextPosX =
> 			pos.x + (this.initialXRelative - this.engine.getModel().getOffsetX() - pos.x) + event.virtualDisplacementX
> 		const linkNextPosY =
> 			pos.y + (this.initialYRelative - this.engine.getModel().getOffsetY() - pos.y) + event.virtualDisplacementY
> 
> 		this.link.getLastPoint().setPosition(linkNextPosX, linkNextPosY)
> ```

It doesn't work for me, it doesn't show the link at all

antonioru commented:
@mathiasbaldissera @dylanvorster 

Ok, I've update the PR by using the relative coordinates (initialXRelative, initialYRelative) instead of the absolute ones and it works for be every time now. 

Please Mathias let me know if you find any other problem, the code I've updated is the following:

```
	/**
	 * Calculates the link's far-end point position on mouse move.
	 * In order to be as precise as possible the mouse initialXRelative & initialYRelative are taken into account.
	 */
	fireMouseMoved(event: AbstractDisplacementStateEvent): any {
		const pos = this.port.getPosition();
		const linkNextPosX = pos.x + (this.initialXRelative - pos.x) + event.virtualDisplacementX;
		const linkNextPosY = pos.y + (this.initialYRelative - pos.y) + event.virtualDisplacementY;

		this.link.getLastPoint().setPosition(linkNextPosX, linkNextPosY);
		this.engine.repaintCanvas();
	}
```

I hope it helps :)

mathiasbaldissera commented:
> let me k

If u use only the relative coordinates, and move the canvas (generating an offsetX and offsetY to the canvas) it'll bug, did u try it?

antonioru commented:
> > let me k
> 
> If u use only the relative coordinates, and move the canvas (generating an offsetX and offsetY to the canvas) it'll bug, did u try it?

I see your point, so the engine offset must be taken into account as well ... 
let me dig a bit more into it... I really can't understand why using the engine offset screws the whole thing in demos :\

antonioru commented:
@mathiasbaldissera 

this seems to finally work even when you move the canvas :)

```
	/**
	 * Calculates the link's far-end point position on mouse move.
	 * In order to be as precise as possible the mouse initialXRelative & initialYRelative are taken into account.
	 */
	fireMouseMoved(event: AbstractDisplacementStateEvent): any {
		const pos = this.port.getPosition();
		const engineOffsetX = this.engine.getModel().getOffsetX();
		const engineOffsetY = this.engine.getModel().getOffsetY();
		const linkNextPosX = pos.x - engineOffsetX + (this.initialXRelative - pos.x) + event.virtualDisplacementX;
		const linkNextPosY = pos.y - engineOffsetY + (this.initialYRelative - pos.y) + event.virtualDisplacementY;

		this.link.getLastPoint().setPosition(linkNextPosX, linkNextPosY);
		this.engine.repaintCanvas();
	}
```

mathiasbaldissera commented:
> @mathiasbaldissera
> 
> this seems to finally work even when you move the canvas :)
> 
> ```
> 	/**
> 	 * Calculates the link's far-end point position on mouse move.
> 	 * In order to be as precise as possible the mouse initialXRelative & initialYRelative are taken into account.
> 	 */
> 	fireMouseMoved(event: AbstractDisplacementStateEvent): any {
> 		const pos = this.port.getPosition();
> 		const engineOffsetX = this.engine.getModel().getOffsetX();
> 		const engineOffsetY = this.engine.getModel().getOffsetY();
> 		const linkNextPosX = pos.x - engineOffsetX + (this.initialXRelative - pos.x) + event.virtualDisplacementX;
> 		const linkNextPosY = pos.y - engineOffsetY + (this.initialYRelative - pos.y) + event.virtualDisplacementY;
> 
> 		this.link.getLastPoint().setPosition(linkNextPosX, linkNextPosY);
> 		this.engine.repaintCanvas();
> 	}
> ```

I'll try it at the afternoon  (now i'ts 11:12 here :v)

antonioru commented:
hehe, take your time mate, I'll update the PR in the meanwhile :) 
(( It's 14:14 here, greetings from London :))

mathiasbaldissera commented:
@antonioru 1 question. How did you exactly run the demos? Every time I try to run i got erros, and I need to test the changes into my own projects

antonioru commented:
> @antonioru 1 question. How did you exactly run the demos? Every time I try to run i got erros, and I need to test the changes into my own projects (I didn't find you on Facebook to ask this question hehehe :v)

1. I git cloned the repository locally on my computer 
2. then `cd react-diagrams` (or the name you gave to your local repository)
3. `npm install` or `yarn install`
4. then `cd packages/diagrams-demo-gallery`
5.  again `npm install` or `yarn install`
6. `npm start` or `yarn start`

i think it's all reported into the doc somewhere

mathiasbaldissera commented:
> > @antonioru 1 question. How did you exactly run the demos? Every time I try to run i got erros, and I need to test the changes into my own projects (I didn't find you on Facebook to ask this question hehehe :v)
> 
> 1. I git cloned the repository locally on my computer
> 2. then `cd react-diagrams` (or the name you gave to your local repository)
> 3. `npm install` or `yarn install`
> 4. then `cd packages/diagrams-demo-gallery`
> 5. again `npm install` or `yarn install`
> 6. `npm start` or `yarn start`
> 
> i think it's all reported into the doc somewhere

I really forgot to run the yarn install in the demo-gallery :man_facepalming:

mathiasbaldissera commented:
> hehe, take your time mate, I'll update the PR in the meanwhile :)
> (( It's 14:14 here, greetings from London :))

It work perfectly now ;)
((now 13:56, greetings from Brazil :smile: )

antonioru commented:
So it just needs to be merged :)) 

Thank you @mathiasbaldissera, you really helped me out 

@dylanvorster

dylanvorster commented:
what a journey xD
